### ISSUE TITLE
Gitea should set `Cache-Control: public` (or unset) on raw files in public repositories so reverse proxies can cache properly.

### ISSUE THREAD

**Issue by ledlamp** [go-gitea/gitea#21391]
### Description

So it appears that Gitea [always sets](https://github.com/go-gitea/gitea/blob/14178c56bb0fbc8b0b5820539e7681a7defeff47/modules/httpcache/httpcache.go#L27) `Cache-Control: private, max-age=300` on raw files, even if the asset is in a public repository and ought to be cachable by a reverse proxy.

This means that if you want to cache gitea right now, you have to override this, but then you'll expose files in private repositories since the proxy can't tell the difference.

It also appears that Gitea disables caching on HTML renders (`no-store`), which makes sense, but it might be a good idea to allow caching of those for unauthenticated visitors, since some views may not change frequently, but could have heavy traffic from unauthenticated visitors (think wikis and readmes with embedded images, etc).

Also, why does Gitea _always_ set a cookie, even if you're not logged in? This is also problematic for caching; Cloudflare won't cache any response with a cookie for example. And since it appears the same `i_like_gitea` cookie is used for both guest sessions and logged in users, there is no way for a custom reverse proxy to tell the difference between visitors and logged-in users.

To the eyes of a reverse proxy, with these two issues, everyone is logged in, and nothing is cacheable!

### Gitea Version

1.17.2, 1.18.0+dev-518-ga813c9d8f

### Can you reproduce the bug on the Gitea demo site?

Yes

### How are you running Gitea?

N/A; checked response headers of raw file responses on gitea.moe and try.gitea.io.

### Database

SQLite

**Comment by delvh** [comment]
Erm… I don't think it is a good idea to cache **user-specific** responses on a CDN or reverse-proxy.
Basically, whenever you request page content, it will be customized for the current user.
This should even be true for API requests as they need to validate that the user is allowed see the response.

**Comment by asdfisdffdg** [comment]
> Erm… I don't think it is a good idea to cache **user-specific** responses on a CDN or reverse-proxy. Basically, whenever you request page content, it will be customized for the current user. This should even be true for API requests as they need to validate that the user is allowed see the response.

Of course, but the files in a **public** repository are _not_ "user-specific" responses. And being able to cache them would be very helpful.

**Comment by asdfisdffdg** [comment]
Things that should be cachable (Cache-Control: public or not set, and no cookies):
- Raw files/assets/media in public repositories
- HTML renders of not-logged-in visitors (even if the cache should be short-lived)

Things that must NOT be cachable (Cache-Control: private, cookies required):
- Raw files/etc in private repositories
- Any HTML render for a logged-in user

Gitea is an ideal candidate for caching/CDN since you might use it to publish projects/wikis/files etc to the general public (so most views would be from unregistered visitors). But putting it behind Cloudflare won't do anything, due to the improper use of cache-control and cookies.

**Comment by wxiaoguang** [comment]
Theoretically, yes

In fact, it's very difficult to make every cache header that correct at the moment. It needs a lot of work, needs to be fine-tuned and tested very carefully.

**Comment by delvh** [comment]
And the problem I can see:
Even on public repos, you still have user-specific content, like the notifications, who is logged in, …
The only solution where this would work is for unauthenticated users as the response looks always the same then.
But that's **a lot** of work for probably minimal gain.
How often do you browse a Gitea instance without being logged in?

**Comment by silverwind** [comment]
HTML content should never be cached by Gitea itself, even for unauthenticated users. Things like relative timestamps will change. Of course reverse proxies can override those cache headers if they so desire, but they will be serving stale content.

**Comment by asdfisdffdg** [comment]
> HTML content should never be cached by Gitea itself, even for unauthenticated users. Things like relative timestamps will change. Of course reverse proxies can override those cache headers if they so desire, but they will be serving stale content.

I never said Gitea itself should cache anything. Gitea should just follow the standard and provide correct semantic information for the proxy. You can't make a proxy cache anything without leaking private files because gitea says everything is private!

And even a few second cache on html could be helpful against DoS. It is up to the webmaster to decide how they want to cache.


> How often do you browse a Gitea instance without being logged in?

Don't tell me I'm the only one who uses Gitea to publish stuff to the public. 90% of requests are unauthenticated visitors.

**Comment by asdfisdffdg** [comment]
> Theoretically, yes
> 
> In fact, it's very difficult to make every cache header that correct at the moment. It needs a lot of work, needs to be fine-tuned and tested very carefully.

if ctx.Repo.Repository.IsPrivate is false, remove "private" directive. You could make private default but add an argument to [AddCacheControlToHeader](https://github.com/go-gitea/gitea/blob/14178c56bb0fbc8b0b5820539e7681a7defeff47/modules/httpcache/httpcache.go#L20) to explicitly make it public, then add that logic [in the thing that serves the raw files](https://github.com/go-gitea/gitea/blob/eaf653f843cb887bc2f4ff664010b4e969dd5446/routers/common/repo.go#L56).

**Comment by shyim** [comment]
How do you want to do then cache invalidation? When you link to a branch and the branch gets updated still the old content is maybe shown as the response was cache-control public

**Comment by Saklad5** [comment]
User-specific content should use `Cache-Control: private`. You almost never need to set this explicitly, as the [default behavior](https://www.rfc-editor.org/rfc/rfc9111.html#section-3) is to treat a response to a request with an `Authorization` header as private unless explicitly marked public.

Conversely, any content that is not user-specific or otherwise private despite being requested with an `Authorization` header should use `Cache-Control: public`.

Every single response should have a `Cache-Control` header and ideally an `ETag`. Anything static with some form of versioning in the URI, particularly with commit hashes, should use `Cache-Control: immutable`.

`Cache-Control: no-store` should be used *very* sparingly. Something unsafe to reuse is almost always better off with `Cache-Control: must-revalidate`, though even that should only be specified for a good reason. With static files, I personally tend to default to `Cache-Control: max-age=86400;stale-while-revalidate=518400`.

Keep in mind that most user-agents will set `Cache-Control: no-cache` in the request if the user manually reloads a page.

Gitea should not worry about actually caching things. However, only it is positioned to intelligently determine what is safe to cache, and it needs to pass that along.

**Comment by Saklad5** [comment]
> How do you want to do then cache invalidation? When you link to a branch and the branch gets updated still the old content is maybe shown as the response was cache-control public

Reuse a response without checking within a brief window, reuse a response while checking in a less brief window.

For branches, I feel the following would be a reasonable, if cautious, balance:
```HTTP
Cache-Control: max-age=60;stale-while-revalidate=240
```
A client will never receive a response older than five minutes, and will quietly check it isn't outdated at least once a minute. Clients can opt out of the background refresh with `Cache-Control: max-stale=0` (meaning it waits for the refresh to finish) or force revalidation every time with `Cache-Control: no-cache`.

**Comment by Saklad5** [comment]
> HTML content should never be cached by Gitea itself, even for unauthenticated users. Things like relative timestamps will change. Of course reverse proxies can override those cache headers if they so desire, but they will be serving stale content.

This doesn't seem like a worthwhile tradeoff. Couldn't Gitea produce relative timestamps with client-side JavaScript, using absolute timestamps in its absence?

**Comment by silverwind** [comment]
> Couldn't Gitea produce relative timestamps with client-side JavaScript

That's already being done IIRC, but still I would not risk caching HTML content, this is far too risky imho.

`Cache-Control: immutable` is worthwile to have for versioned assets like JS/CSS.
`Cache-Control: public` I guess it's possible to have this for JS/CSS as well, given that there Gitea source code is open source.

**Comment by Saklad5** [comment]
> That's already being done IIRC, but still I would not risk caching HTML content, this is far too risky imho.

What do you mean, "risky"?

**Comment by silverwind** [comment]
Risky in the sense that dynamic content (based on server state and HTTP headers) will go stale and the user in turn getting a outdated page. Take for example the language setting, with cached HTML, user would never see a language change until hard reload.

**Comment by Saklad5** [comment]
> Risky in the sense that dynamic content (based on server state and HTTP headers) will go stale and the user in turn getting a outdated page. Take for example the language setting, with cached HTML, user would never see a language change until hard page reload.

I recommend reading [RFC 9111](https://www.rfc-editor.org/rfc/rfc9111). HTTP caching uses the `Vary` header to designate which request headers affect the applicability of the cached response. In that example, `Accept-Language` would be there.

**Comment by silverwind** [comment]
Language is server state stored in a cookie, so we can't use `Vary`. If at all, `Accept-Language` is only used for the initial value.

**Comment by Saklad5** [comment]
It'd be better than nothing, but not ideal. Why store language there?

**Comment by silverwind** [comment]
Well, it's an easy way to allow users to switch language, e.g. they override what their browser has configured.

This is only one example that shows that HTML can not be made cacheable, there are certainly more such cases. And even if we do it, only logged-out users would benefit, but that is not the primary use case where users are usually logged in.

**Comment by Saklad5** [comment]
The important bit is browser caching, not a shared cache.

**Comment by silverwind** [comment]
What is the difference between `max-age=0, private, must-revalidate` and `no-store`? In both cases, the browser must retrieve the data again, in the first case it can cache, but what is the benefit if that cache is never used?

**Comment by Saklad5** [comment]
The browser does not need to retrieve the data again in the first case, necessarily. The server can tell it that the cached content is fine, and the browser can use that without having to download it again.

Again, you should read RFC 9111.

**Comment by Saklad5** [comment]
At the very least, RSS feeds **need** to be aggressively cached, as the entire specification revolves around that. It is completely unacceptable to have feed readers download the entire feed every time they check for changes.

**Comment by silverwind** [comment]
Agree, we can certainly improve the caching headers for content like RSS and versioned assets. Unversioned assets (not sure we have any) and especially HTML should only be cached with caution.

**Comment by Saklad5** [comment]
There is literally no risk here: `max-age=0, must-revalidate` means cached content is only used after checking to make sure it isn't outdated. The server explicitly approves every use with 304 responses.

Again, read the specification.

**Comment by ledlamp** [comment]
FWIW I had to block an entire country from my gitea for using it as a static file host and overloading it which wouldn't have been a problem if cloudflare could just cache the raw files at least

**Comment by wxiaoguang** [comment]
Will be fixed by #33861
